tests/marray/marray_operators.h

```diff
@@ -0,0 +1,491 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Copyright (c) 2022 The Khronos Group Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+*******************************************************************************/
+
+#ifndef SYCLCTS_TESTS_MARRAY_MARRAY_OPERATOR_H
+#define SYCLCTS_TESTS_MARRAY_MARRAY_OPERATOR_H
+
+#include "../common/common.h"
+#include "../common/disabled_for_test_case.h"
+#include "../common/section_name_builder.h"
+#include "marray_common.h"
+#include "marray_operator_helper.h"
+
+#include <valarray>
+
+namespace marray_operators {
+
+template <typename DataT, typename NumElementsT>
+struct operators_helper {
+  static constexpr std::size_t NumElements = NumElementsT::value;
+  using marray_t = sycl::marray<DataT, NumElements>;
+  using varray_t = std::valarray<DataT>;
+
+  template <typename init_func, typename array_type>
+  static void init(array_type &ma) {
+    for (std::size_t i = 0; i < NumElements; i++) {
+      ma[i] = init_func::template init<DataT>(i);
+    }
+  }
+
+  template <typename init_func>
+  static void init(DataT &d) {
+    d = init_func::template init<DataT>();
+  }
+};
+
+/**
+ * @brief Define several sequences to initialize array instances. */
+
+struct seq_inc {
+  template <typename T>
+  static T init(std::size_t i) {
+    return T(i + 1);
+  }
+};
+
+template <std::size_t NumElements>
+struct seq_dec {
+  template <typename T>
+  static T init(std::size_t i) {
+    return T(NumElements - i);
+  }
+};
+
+struct seq_one {
+  template <typename T>
+  static T init(std::size_t) {
+    return T(1);
```

Comment:
Is this because the constructor is explicit?

Reply:
I have replaced this with `{1}`. Note that `T` here is the data type, not the math array type. I have updated the template parameter name to reflect this.

Note that it is not possible to do for `T(NumElements - i);` as it cannot convert the `size_t` non-constant expression `NumElements - i` to some types e.g. `double`. Nor is it possible to replace `T(2)` by `{2}` because `2` cannot be narrowed to `bool` at compile time.